<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>vg: vg::PhasedGenome Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">vg
   </div>
   <div id="projectbrief">tools for working with variation graphs</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacevg.html">vg</a></li><li class="navelem"><a class="el" href="classvg_1_1PhasedGenome.html">PhasedGenome</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#nested-classes">Classes</a> &#124;
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="classvg_1_1PhasedGenome-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">vg::PhasedGenome Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;phased_genome.hpp&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="nested-classes"></a>
Classes</h2></td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">struct &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:"><td class="memItemLeft" align="right" valign="top">class &#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome_1_1iterator.html">iterator</a></td></tr>
<tr class="separator:"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a3360afd88251aedab3c18136835893b9"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a3360afd88251aedab3c18136835893b9">PhasedGenome</a> (<a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;<a class="el" href="classvg_1_1PhasedGenome.html#a3da5446fd6bafdab3e4b8532a9dbe18b">snarl_manager</a>)</td></tr>
<tr class="memdesc:a3360afd88251aedab3c18136835893b9"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classvg_1_1Constructor.html">Constructor</a>.  <a href="#a3360afd88251aedab3c18136835893b9">More...</a><br/></td></tr>
<tr class="separator:a3360afd88251aedab3c18136835893b9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a08609371573b24264b5f5e14af6d0139"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a08609371573b24264b5f5e14af6d0139">~PhasedGenome</a> ()</td></tr>
<tr class="separator:a08609371573b24264b5f5e14af6d0139"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad36e2c71ff5b2714d1fe1bb6af075f9c"><td class="memTemplParams" colspan="2">template&lt;typename NodeTraversalIterator &gt; </td></tr>
<tr class="memitem:ad36e2c71ff5b2714d1fe1bb6af075f9c"><td class="memTemplItemLeft" align="right" valign="top">int&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#ad36e2c71ff5b2714d1fe1bb6af075f9c">add_haplotype</a> (NodeTraversalIterator first, NodeTraversalIterator last)</td></tr>
<tr class="separator:ad36e2c71ff5b2714d1fe1bb6af075f9c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a72e27ed72d346717ce196d5b2081365b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a72e27ed72d346717ce196d5b2081365b">build_indices</a> ()</td></tr>
<tr class="separator:a72e27ed72d346717ce196d5b2081365b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af17e5f99a07028d41f94885480cdddf7"><td class="memItemLeft" align="right" valign="top">size_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#af17e5f99a07028d41f94885480cdddf7">num_haplotypes</a> ()</td></tr>
<tr class="separator:af17e5f99a07028d41f94885480cdddf7"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5ec1fda02f9b639fa315337290dca80f"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1PhasedGenome_1_1iterator.html">iterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a5ec1fda02f9b639fa315337290dca80f">begin</a> (int which_haplotype)</td></tr>
<tr class="separator:a5ec1fda02f9b639fa315337290dca80f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a621ecd8fc372256aa037a8beb9ea4c39"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1PhasedGenome_1_1iterator.html">iterator</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a621ecd8fc372256aa037a8beb9ea4c39">end</a> (int which_haplotype)</td></tr>
<tr class="separator:a621ecd8fc372256aa037a8beb9ea4c39"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae325c8488e737b75b980796aacf37246"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#ae325c8488e737b75b980796aacf37246">swap_alleles</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;site, int haplotype_1, int haplotype_2)</td></tr>
<tr class="separator:ae325c8488e737b75b980796aacf37246"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad0fdafbc6cb6fbfb576be92f6f6bc17d"><td class="memTemplParams" colspan="2">template&lt;typename NodeTraversalIterator &gt; </td></tr>
<tr class="memitem:ad0fdafbc6cb6fbfb576be92f6f6bc17d"><td class="memTemplItemLeft" align="right" valign="top">void&#160;</td><td class="memTemplItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#ad0fdafbc6cb6fbfb576be92f6f6bc17d">set_allele</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;site, NodeTraversalIterator first, NodeTraversalIterator last, int which_haplotype)</td></tr>
<tr class="separator:ad0fdafbc6cb6fbfb576be92f6f6bc17d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a6f89ab856a31dfb7258fd7487c051a5a"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a6f89ab856a31dfb7258fd7487c051a5a">get_allele</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;site, int which_haplotype)</td></tr>
<tr class="separator:a6f89ab856a31dfb7258fd7487c051a5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aab26653dddeea64012b2629b4ebc3836"><td class="memItemLeft" align="right" valign="top">int32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#aab26653dddeea64012b2629b4ebc3836">optimal_score_on_genome</a> (const <a class="el" href="structvg_1_1MultipathAlignment.html">MultipathAlignment</a> &amp;multipath_aln, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph)</td></tr>
<tr class="separator:aab26653dddeea64012b2629b4ebc3836"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:ab7badd2e4e135f118fc14f7e0aad1b5a"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#ab7badd2e4e135f118fc14f7e0aad1b5a">build_site_indices_internal</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *snarl)</td></tr>
<tr class="separator:ab7badd2e4e135f118fc14f7e0aad1b5a"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a02e624546cf3f5675fbfa128c080c75d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a02e624546cf3f5675fbfa128c080c75d">insert_left</a> (<a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> node_traversal, <a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> *haplo_node)</td></tr>
<tr class="memdesc:a02e624546cf3f5675fbfa128c080c75d"><td class="mdescLeft">&#160;</td><td class="mdescRight">Insert a node traversal to the left of this haplotype node and update indices.  <a href="#a02e624546cf3f5675fbfa128c080c75d">More...</a><br/></td></tr>
<tr class="separator:a02e624546cf3f5675fbfa128c080c75d"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:adfbd6242bff1445e5210f4046150fc85"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#adfbd6242bff1445e5210f4046150fc85">insert_right</a> (<a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> node_traversal, <a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> *haplo_node)</td></tr>
<tr class="memdesc:adfbd6242bff1445e5210f4046150fc85"><td class="mdescLeft">&#160;</td><td class="mdescRight">Insert a node traversal to the right of this haplotype node and update indices.  <a href="#adfbd6242bff1445e5210f4046150fc85">More...</a><br/></td></tr>
<tr class="separator:adfbd6242bff1445e5210f4046150fc85"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a3706e208421ac10d5695ed5d5ec45090"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a3706e208421ac10d5695ed5d5ec45090">remove</a> (<a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> *haplo_node)</td></tr>
<tr class="memdesc:a3706e208421ac10d5695ed5d5ec45090"><td class="mdescLeft">&#160;</td><td class="mdescRight">Remove this haplotype node from its haplotype and update indices.  <a href="#a3706e208421ac10d5695ed5d5ec45090">More...</a><br/></td></tr>
<tr class="separator:a3706e208421ac10d5695ed5d5ec45090"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a181b4e71f8debc78ceccf476ce36abcd"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a181b4e71f8debc78ceccf476ce36abcd">swap_label</a> (const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;site, <a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a> &amp;haplotype_1, <a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a> &amp;haplotype_2)</td></tr>
<tr class="memdesc:a181b4e71f8debc78ceccf476ce36abcd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Update a subsite's location in indices after swapping its parent allele.  <a href="#a181b4e71f8debc78ceccf476ce36abcd">More...</a><br/></td></tr>
<tr class="separator:a181b4e71f8debc78ceccf476ce36abcd"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a3da5446fd6bafdab3e4b8532a9dbe18b"><td class="memItemLeft" align="right" valign="top"><a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a3da5446fd6bafdab3e4b8532a9dbe18b">snarl_manager</a></td></tr>
<tr class="separator:a3da5446fd6bafdab3e4b8532a9dbe18b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa7b2bebc28cc37da09216fca8361c75c"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#aa7b2bebc28cc37da09216fca8361c75c">haplotypes</a></td></tr>
<tr class="memdesc:aa7b2bebc28cc37da09216fca8361c75c"><td class="mdescLeft">&#160;</td><td class="mdescRight">All haplotypes in the genome (generally 2 per chromosome)  <a href="#aa7b2bebc28cc37da09216fca8361c75c">More...</a><br/></td></tr>
<tr class="separator:aa7b2bebc28cc37da09216fca8361c75c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a00b61ff3ca8566865774b0661d5f47c6"><td class="memItemLeft" align="right" valign="top">unordered_map&lt; int64_t, list<br class="typebreak"/>
&lt; <a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> * &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#a00b61ff3ca8566865774b0661d5f47c6">node_locations</a></td></tr>
<tr class="memdesc:a00b61ff3ca8566865774b0661d5f47c6"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classvg_1_1Index.html">Index</a> of where nodes from the graph occur in the phased genome.  <a href="#a00b61ff3ca8566865774b0661d5f47c6">More...</a><br/></td></tr>
<tr class="separator:a00b61ff3ca8566865774b0661d5f47c6"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:accfaae53e12f924b914623734a68102f"><td class="memItemLeft" align="right" valign="top">unordered_map&lt; int64_t, const <br class="typebreak"/>
<a class="el" href="structvg_1_1Snarl.html">Snarl</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#accfaae53e12f924b914623734a68102f">site_starts</a></td></tr>
<tr class="memdesc:accfaae53e12f924b914623734a68102f"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classvg_1_1Index.html">Index</a> of which nodes are starts of Snarls.  <a href="#accfaae53e12f924b914623734a68102f">More...</a><br/></td></tr>
<tr class="separator:accfaae53e12f924b914623734a68102f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc338c2bf1776ec1ccb29c8efee6aaa3"><td class="memItemLeft" align="right" valign="top">unordered_map&lt; int64_t, const <br class="typebreak"/>
<a class="el" href="structvg_1_1Snarl.html">Snarl</a> * &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1PhasedGenome.html#afc338c2bf1776ec1ccb29c8efee6aaa3">site_ends</a></td></tr>
<tr class="memdesc:afc338c2bf1776ec1ccb29c8efee6aaa3"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classvg_1_1Index.html">Index</a> of which nodes are ends of Snarls.  <a href="#afc338c2bf1776ec1ccb29c8efee6aaa3">More...</a><br/></td></tr>
<tr class="separator:afc338c2bf1776ec1ccb29c8efee6aaa3"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>A collection of haplotypes that represent all of the chromosomes of a genome (including phasing) as walks through a variation graph. Designed for fast editing at a site level, so it maintains indices of sites for that purpose. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a3360afd88251aedab3c18136835893b9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vg::PhasedGenome::PhasedGenome </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a> &amp;&#160;</td>
          <td class="paramname"><em>snarl_manager</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p><a class="el" href="classvg_1_1Constructor.html">Constructor</a>. </p>

</div>
</div>
<a class="anchor" id="a08609371573b24264b5f5e14af6d0139"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vg::PhasedGenome::~PhasedGenome </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="ad36e2c71ff5b2714d1fe1bb6af075f9c"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename NodeTraversalIterator &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">int vg::PhasedGenome::add_haplotype </td>
          <td>(</td>
          <td class="paramtype">NodeTraversalIterator&#160;</td>
          <td class="paramname"><em>first</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">NodeTraversalIterator&#160;</td>
          <td class="paramname"><em>last</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Build a haplotype in place from an iterator that returns <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> objects from its dereference operator (allows construction without instantiating the haplotype elsewhere) returns the numerical id of the new haplotype</p>
<p>note: the haplotype must have at least one node </p>

</div>
</div>
<a class="anchor" id="a5ec1fda02f9b639fa315337290dca80f"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1PhasedGenome_1_1iterator.html">PhasedGenome::iterator</a> vg::PhasedGenome::begin </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>which_haplotype</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Unidirectional iterator starting at the left telomere and moving to the right. Requires a haplotype id as input </p>

</div>
</div>
<a class="anchor" id="a72e27ed72d346717ce196d5b2081365b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::build_indices </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Construct the site ends, node locations, and haplotype site location indices. This method is intended to be called one time after building haplotypes. After this, the are maintained automatically during edit operations. </p>

</div>
</div>
<a class="anchor" id="ab7badd2e4e135f118fc14f7e0aad1b5a"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::build_site_indices_internal </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> *&#160;</td>
          <td class="paramname"><em>snarl</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="a621ecd8fc372256aa037a8beb9ea4c39"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1PhasedGenome_1_1iterator.html">PhasedGenome::iterator</a> vg::PhasedGenome::end </td>
          <td>(</td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>which_haplotype</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Iterator representing the past-the-last position of the given haplotype, with the last position being the right telomere node. </p>

</div>
</div>
<a class="anchor" id="a6f89ab856a31dfb7258fd7487c051a5a"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; <a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a> &gt; vg::PhasedGenome::get_allele </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;&#160;</td>
          <td class="paramname"><em>site</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>which_haplotype</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Returns a vector of node traversals representing the allele at the indicated site on the indicated haplotype. The allele is returned in the orientation of the <a class="el" href="structvg_1_1Snarl.html" title="Describes a subgraph that is connected to the rest of the graph by two nodes. ">Snarl</a> (start to end), not necessarily the orientation on the haplotype. The start and end nodes of the <a class="el" href="structvg_1_1Snarl.html" title="Describes a subgraph that is connected to the rest of the graph by two nodes. ">Snarl</a> are not included in the allele. </p>

</div>
</div>
<a class="anchor" id="a02e624546cf3f5675fbfa128c080c75d"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::insert_left </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a>&#160;</td>
          <td class="paramname"><em>node_traversal</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> *&#160;</td>
          <td class="paramname"><em>haplo_node</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Insert a node traversal to the left of this haplotype node and update indices. </p>

</div>
</div>
<a class="anchor" id="adfbd6242bff1445e5210f4046150fc85"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::insert_right </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1NodeTraversal.html">NodeTraversal</a>&#160;</td>
          <td class="paramname"><em>node_traversal</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> *&#160;</td>
          <td class="paramname"><em>haplo_node</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Insert a node traversal to the right of this haplotype node and update indices. </p>

</div>
</div>
<a class="anchor" id="af17e5f99a07028d41f94885480cdddf7"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">size_t vg::PhasedGenome::num_haplotypes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

</div>
</div>
<a class="anchor" id="aab26653dddeea64012b2629b4ebc3836"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32_t vg::PhasedGenome::optimal_score_on_genome </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1MultipathAlignment.html">MultipathAlignment</a> &amp;&#160;</td>
          <td class="paramname"><em>multipath_aln</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Returns the score of the highest scoring alignment contained in the multipath alignment that is restricted to the phased genome's paths through the variation graph.</p>
<p>Note: assumes that <a class="el" href="structvg_1_1MultipathAlignment.html">MultipathAlignment</a> has 'start' field filled in </p>

</div>
</div>
<a class="anchor" id="a3706e208421ac10d5695ed5d5ec45090"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::remove </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a> *&#160;</td>
          <td class="paramname"><em>haplo_node</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Remove this haplotype node from its haplotype and update indices. </p>

</div>
</div>
<a class="anchor" id="ad0fdafbc6cb6fbfb576be92f6f6bc17d"></a>
<div class="memitem">
<div class="memproto">
<div class="memtemplate">
template&lt;typename NodeTraversalIterator &gt; </div>
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::set_allele </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;&#160;</td>
          <td class="paramname"><em>site</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">NodeTraversalIterator&#160;</td>
          <td class="paramname"><em>first</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">NodeTraversalIterator&#160;</td>
          <td class="paramname"><em>last</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>which_haplotype</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Set the allele at a site with an iterator that yields its node sequence. The allele should be provided in the order indicated by the <a class="el" href="structvg_1_1Snarl.html" title="Describes a subgraph that is connected to the rest of the graph by two nodes. ">Snarl</a> (i.e. from start to end) and it should not include the boundary nodes of the <a class="el" href="structvg_1_1Snarl.html" title="Describes a subgraph that is connected to the rest of the graph by two nodes. ">Snarl</a>.</p>
<p>Note: This method does not check that the allele path takes only edges that are actually included in the graph, so client must ensure this itself. </p>

</div>
</div>
<a class="anchor" id="ae325c8488e737b75b980796aacf37246"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::swap_alleles </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;&#160;</td>
          <td class="paramname"><em>site</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>haplotype_1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>haplotype_2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Swap the allele from between two haplotypes, maintaining all indices. If a nested site is being swapped, this method should be called only once for the top-most site. Child sites are swapped along with the top-most site automatically. </p>

</div>
</div>
<a class="anchor" id="a181b4e71f8debc78ceccf476ce36abcd"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::PhasedGenome::swap_label </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Snarl.html">Snarl</a> &amp;&#160;</td>
          <td class="paramname"><em>site</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a> &amp;&#160;</td>
          <td class="paramname"><em>haplotype_1</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a> &amp;&#160;</td>
          <td class="paramname"><em>haplotype_2</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Update a subsite's location in indices after swapping its parent allele. </p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="aa7b2bebc28cc37da09216fca8361c75c"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="classvg_1_1PhasedGenome_1_1Haplotype.html">Haplotype</a>*&gt; vg::PhasedGenome::haplotypes</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>All haplotypes in the genome (generally 2 per chromosome) </p>

</div>
</div>
<a class="anchor" id="a00b61ff3ca8566865774b0661d5f47c6"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">unordered_map&lt;int64_t, list&lt;<a class="el" href="structvg_1_1PhasedGenome_1_1HaplotypeNode.html">HaplotypeNode</a>*&gt; &gt; vg::PhasedGenome::node_locations</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><a class="el" href="classvg_1_1Index.html">Index</a> of where nodes from the graph occur in the phased genome. </p>

</div>
</div>
<a class="anchor" id="afc338c2bf1776ec1ccb29c8efee6aaa3"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">unordered_map&lt;int64_t, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a>*&gt; vg::PhasedGenome::site_ends</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><a class="el" href="classvg_1_1Index.html">Index</a> of which nodes are ends of Snarls. </p>

</div>
</div>
<a class="anchor" id="accfaae53e12f924b914623734a68102f"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">unordered_map&lt;int64_t, const <a class="el" href="structvg_1_1Snarl.html">Snarl</a>*&gt; vg::PhasedGenome::site_starts</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p><a class="el" href="classvg_1_1Index.html">Index</a> of which nodes are starts of Snarls. </p>

</div>
</div>
<a class="anchor" id="a3da5446fd6bafdab3e4b8532a9dbe18b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname"><a class="el" href="classvg_1_1SnarlManager.html">SnarlManager</a>&amp; vg::PhasedGenome::snarl_manager</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="phased__genome_8hpp.html">phased_genome.hpp</a></li>
<li>src/<a class="el" href="phased__genome_8cpp.html">phased_genome.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Mon Oct 1 2018 17:57:07 for vg by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.6
</small></address>
</body>
</html>
