<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>vg: vg::Transcriptome Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">vg
   </div>
   <div id="projectbrief">tools for working with variation graphs</div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="hierarchy.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(5)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(6)"><span class="SelectionMark">&#160;</span>Typedefs</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(7)"><span class="SelectionMark">&#160;</span>Enumerations</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(8)"><span class="SelectionMark">&#160;</span>Enumerator</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(9)"><span class="SelectionMark">&#160;</span>Friends</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(10)"><span class="SelectionMark">&#160;</span>Macros</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div id="nav-path" class="navpath">
  <ul>
<li class="navelem"><a class="el" href="namespacevg.html">vg</a></li><li class="navelem"><a class="el" href="classvg_1_1Transcriptome.html">Transcriptome</a></li>  </ul>
</div>
</div><!-- top -->
<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="classvg_1_1Transcriptome-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">vg::Transcriptome Class Reference</div>  </div>
</div><!--header-->
<div class="contents">

<p><code>#include &lt;transcriptome.hpp&gt;</code></p>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a28cf9247552f4c54c2852a63effb588e"><td class="memItemLeft" align="right" valign="top">&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a28cf9247552f4c54c2852a63effb588e">Transcriptome</a> ()</td></tr>
<tr class="separator:a28cf9247552f4c54c2852a63effb588e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a9b8157c2b1b8188363d3b02c5dca1775"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a9b8157c2b1b8188363d3b02c5dca1775">add_transcripts</a> (istream &amp;transcript_stream, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const gbwt::GBWT &amp;haplotype_index)</td></tr>
<tr class="separator:a9b8157c2b1b8188363d3b02c5dca1775"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a79f5fde9d7133ce828600482705c985e"><td class="memItemLeft" align="right" valign="top">const vector&lt; <a class="el" href="structvg_1_1Path.html">Path</a> &gt; &amp;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a79f5fde9d7133ce828600482705c985e">transcript_paths</a> () const </td></tr>
<tr class="memdesc:a79f5fde9d7133ce828600482705c985e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns transcript paths.  <a href="#a79f5fde9d7133ce828600482705c985e">More...</a><br/></td></tr>
<tr class="separator:a79f5fde9d7133ce828600482705c985e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa87de45115c76c3acfcb868391ae97bd"><td class="memItemLeft" align="right" valign="top">int32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#aa87de45115c76c3acfcb868391ae97bd">size</a> () const </td></tr>
<tr class="memdesc:aa87de45115c76c3acfcb868391ae97bd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Returns number of transcript paths.  <a href="#aa87de45115c76c3acfcb868391ae97bd">More...</a><br/></td></tr>
<tr class="separator:aa87de45115c76c3acfcb868391ae97bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af921ea87f47cf03505aecd2286a2c05b"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#af921ea87f47cf03505aecd2286a2c05b">edit_graph</a> (<a class="el" href="classvg_1_1VG.html">VG</a> *graph, const bool add_paths)</td></tr>
<tr class="separator:af921ea87f47cf03505aecd2286a2c05b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad4f4ab33718c5e5eff171f698e6ddf5e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#ad4f4ab33718c5e5eff171f698e6ddf5e">construct_gbwt</a> (gbwt::GBWTBuilder *gbwt_builder) const </td></tr>
<tr class="memdesc:ad4f4ab33718c5e5eff171f698e6ddf5e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Add transcript paths as threads in GBWT index.  <a href="#ad4f4ab33718c5e5eff171f698e6ddf5e">More...</a><br/></td></tr>
<tr class="separator:ad4f4ab33718c5e5eff171f698e6ddf5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ab76604980376271d6bbdffed718d8fe9"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#ab76604980376271d6bbdffed718d8fe9">write_gam_alignments</a> (ostream *gam_ostream) const </td></tr>
<tr class="memdesc:ab76604980376271d6bbdffed718d8fe9"><td class="mdescLeft">&#160;</td><td class="mdescRight">Writes transcript paths as alignments to a gam file.  <a href="#ab76604980376271d6bbdffed718d8fe9">More...</a><br/></td></tr>
<tr class="separator:ab76604980376271d6bbdffed718d8fe9"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a67d0d4546497376c5c31d248721f9c5e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a67d0d4546497376c5c31d248721f9c5e">write_fasta_sequences</a> (ostream *fasta_ostream, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph) const </td></tr>
<tr class="memdesc:a67d0d4546497376c5c31d248721f9c5e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Writes transcript path sequences to a fasta file.  <a href="#a67d0d4546497376c5c31d248721f9c5e">More...</a><br/></td></tr>
<tr class="separator:a67d0d4546497376c5c31d248721f9c5e"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:ac2261cd2e4236c02dbfded216253c923"><td class="memItemLeft" align="right" valign="top">int32_t&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#ac2261cd2e4236c02dbfded216253c923">num_threads</a> = 1</td></tr>
<tr class="memdesc:ac2261cd2e4236c02dbfded216253c923"><td class="mdescLeft">&#160;</td><td class="mdescRight">Number of threads used for transcript path construction.  <a href="#ac2261cd2e4236c02dbfded216253c923">More...</a><br/></td></tr>
<tr class="separator:ac2261cd2e4236c02dbfded216253c923"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a56a00bd350c02600e5474fc89791d5bd"><td class="memItemLeft" align="right" valign="top">string&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a56a00bd350c02600e5474fc89791d5bd">transcript_tag</a> = &quot;transcript_id&quot;</td></tr>
<tr class="memdesc:a56a00bd350c02600e5474fc89791d5bd"><td class="mdescLeft">&#160;</td><td class="mdescRight">Attribute tag used to parse the transcript id/name in the gtf/gff file.  <a href="#a56a00bd350c02600e5474fc89791d5bd">More...</a><br/></td></tr>
<tr class="separator:a56a00bd350c02600e5474fc89791d5bd"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:afc815aefb14bad4cec9aeccdd9872652"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#afc815aefb14bad4cec9aeccdd9872652">use_embedded_paths</a> = false</td></tr>
<tr class="memdesc:afc815aefb14bad4cec9aeccdd9872652"><td class="mdescLeft">&#160;</td><td class="mdescRight">Use paths embedded in the graph for transcript path construction.  <a href="#afc815aefb14bad4cec9aeccdd9872652">More...</a><br/></td></tr>
<tr class="separator:afc815aefb14bad4cec9aeccdd9872652"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaef5359c4785d9e01e989a7875eed85e"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#aaef5359c4785d9e01e989a7875eed85e">collapse_transcript_paths</a> = true</td></tr>
<tr class="memdesc:aaef5359c4785d9e01e989a7875eed85e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Collapse identical transcript paths.  <a href="#aaef5359c4785d9e01e989a7875eed85e">More...</a><br/></td></tr>
<tr class="separator:aaef5359c4785d9e01e989a7875eed85e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a78e782cd6f5d5c38730fbe54cfcbe3fc"><td class="memItemLeft" align="right" valign="top">bool&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a78e782cd6f5d5c38730fbe54cfcbe3fc">filter_reference_transcript_paths</a> = false</td></tr>
<tr class="memdesc:a78e782cd6f5d5c38730fbe54cfcbe3fc"><td class="mdescLeft">&#160;</td><td class="mdescRight"><a class="el" href="classvg_1_1Filter.html">Filter</a> transcript paths originating from a reference chromosome/contig.  <a href="#a78e782cd6f5d5c38730fbe54cfcbe3fc">More...</a><br/></td></tr>
<tr class="separator:a78e782cd6f5d5c38730fbe54cfcbe3fc"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:a4927274f931b545accce009aeb9bd852"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a4927274f931b545accce009aeb9bd852">add_exon</a> (<a class="el" href="structvg_1_1Transcript.html">Transcript</a> *transcript, const pair&lt; int32_t, int32_t &gt; &amp;exon_pos, const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;chrom_path_index) const </td></tr>
<tr class="separator:a4927274f931b545accce009aeb9bd852"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a130ef684755419776eaa83aae6a67bfa"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a130ef684755419776eaa83aae6a67bfa">reorder_exons</a> (<a class="el" href="structvg_1_1Transcript.html">Transcript</a> *transcript) const </td></tr>
<tr class="separator:a130ef684755419776eaa83aae6a67bfa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a963a41a941cd7ac2cf4857b110ed6e10"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a963a41a941cd7ac2cf4857b110ed6e10">project_transcripts</a> (const vector&lt; <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &gt; &amp;transcripts, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const gbwt::GBWT &amp;haplotype_index, const float mean_node_length)</td></tr>
<tr class="separator:a963a41a941cd7ac2cf4857b110ed6e10"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a429527c91bc59be1efe68e1ca5984833"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a429527c91bc59be1efe68e1ca5984833">project_transcripts_callback</a> (const int32_t thread_idx, const vector&lt; <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &gt; &amp;transcripts, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const gbwt::GBWT &amp;haplotype_index, const float mean_node_length)</td></tr>
<tr class="memdesc:a429527c91bc59be1efe68e1ca5984833"><td class="mdescLeft">&#160;</td><td class="mdescRight">Threaded transcript projecting.  <a href="#a429527c91bc59be1efe68e1ca5984833">More...</a><br/></td></tr>
<tr class="separator:a429527c91bc59be1efe68e1ca5984833"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a5671f47e3233708fca381f1dce05b987"><td class="memItemLeft" align="right" valign="top">list&lt; <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a5671f47e3233708fca381f1dce05b987">project_transcript_gbwt</a> (const <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &amp;cur_transcript, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph, const gbwt::GBWT &amp;haplotype_index, const float mean_node_length) const </td></tr>
<tr class="memdesc:a5671f47e3233708fca381f1dce05b987"><td class="mdescLeft">&#160;</td><td class="mdescRight">Projects transcripts onto haplotypes in a GBWT index and returns resulting transcript paths.  <a href="#a5671f47e3233708fca381f1dce05b987">More...</a><br/></td></tr>
<tr class="separator:a5671f47e3233708fca381f1dce05b987"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a4808b2bd9c101b32da4e217abcac988b"><td class="memItemLeft" align="right" valign="top">vector&lt; pair&lt; <a class="el" href="namespacevg.html#a8ae488d495e344881cbcf599ebdd2355">exon_nodes_t</a>, <br class="typebreak"/>
vector&lt; gbwt::size_type &gt; &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a4808b2bd9c101b32da4e217abcac988b">get_exon_haplotypes</a> (const <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">vg::id_t</a> start_node, const <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">vg::id_t</a> end_node, const gbwt::GBWT &amp;haplotype_index, const int32_t expected_length) const </td></tr>
<tr class="separator:a4808b2bd9c101b32da4e217abcac988b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ac6aef1ffe0baa75f948859590e2ff33f"><td class="memItemLeft" align="right" valign="top">list&lt; <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#ac6aef1ffe0baa75f948859590e2ff33f">project_transcript_embedded</a> (const <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &amp;cur_transcript, <a class="el" href="classvg_1_1VG.html">VG</a> &amp;graph) const </td></tr>
<tr class="memdesc:ac6aef1ffe0baa75f948859590e2ff33f"><td class="mdescLeft">&#160;</td><td class="mdescRight">Projects transcripts onto embedded paths in a variation graph and returns resulting transcript paths.  <a href="#ac6aef1ffe0baa75f948859590e2ff33f">More...</a><br/></td></tr>
<tr class="separator:ac6aef1ffe0baa75f948859590e2ff33f"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a0f3b8b6ed1aa9751628294e89aa10e39"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a0f3b8b6ed1aa9751628294e89aa10e39">collapse_identical_paths</a> (list&lt; <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a> &gt; *cur_transcript_paths) const </td></tr>
<tr class="separator:a0f3b8b6ed1aa9751628294e89aa10e39"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a127a44cf9f2a9d01064325e8e68cc948"><td class="memItemLeft" align="right" valign="top">vector&lt; <a class="el" href="structvg_1_1Path.html">Path</a> &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a127a44cf9f2a9d01064325e8e68cc948">_transcriptome</a></td></tr>
<tr class="separator:a127a44cf9f2a9d01064325e8e68cc948"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a7f981a9d8172e931cc7c59b60de62caa"><td class="memItemLeft" align="right" valign="top">mutex&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classvg_1_1Transcriptome.html#a7f981a9d8172e931cc7c59b60de62caa">trancriptome_mutex</a></td></tr>
<tr class="memdesc:a7f981a9d8172e931cc7c59b60de62caa"><td class="mdescLeft">&#160;</td><td class="mdescRight">Mutex used for adding transcript paths to transcriptome.  <a href="#a7f981a9d8172e931cc7c59b60de62caa">More...</a><br/></td></tr>
<tr class="separator:a7f981a9d8172e931cc7c59b60de62caa"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock"><p>Class that defines a transcriptome represented by a set of paths. </p>
</div><h2 class="groupheader">Constructor &amp; Destructor Documentation</h2>
<a class="anchor" id="a28cf9247552f4c54c2852a63effb588e"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vg::Transcriptome::Transcriptome </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">inline</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

</div>
</div>
<h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a4927274f931b545accce009aeb9bd852"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::add_exon </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1Transcript.html">Transcript</a> *&#160;</td>
          <td class="paramname"><em>transcript</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const pair&lt; int32_t, int32_t &gt; &amp;&#160;</td>
          <td class="paramname"><em>exon_pos</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="structvg_1_1PathIndex.html">PathIndex</a> &amp;&#160;</td>
          <td class="paramname"><em>chrom_path_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Finds the position of each end of a exon on a path in the variation graph and adds the exon to a transcript. </p>

</div>
</div>
<a class="anchor" id="a9b8157c2b1b8188363d3b02c5dca1775"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::add_transcripts </td>
          <td>(</td>
          <td class="paramtype">istream &amp;&#160;</td>
          <td class="paramname"><em>transcript_stream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const gbwt::GBWT &amp;&#160;</td>
          <td class="paramname"><em>haplotype_index</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Constructs transcript paths by projecting transcripts from a gtf/gff file onto embedded paths in a variation graph and/or haplotypes in a GBWT index. </p>

</div>
</div>
<a class="anchor" id="a0f3b8b6ed1aa9751628294e89aa10e39"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::collapse_identical_paths </td>
          <td>(</td>
          <td class="paramtype">list&lt; <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a> &gt; *&#160;</td>
          <td class="paramname"><em>cur_transcript_paths</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Collapses identical transcript paths. The number of collapsed copies are recorded in the data structure <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a>. </p>

</div>
</div>
<a class="anchor" id="ad4f4ab33718c5e5eff171f698e6ddf5e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::construct_gbwt </td>
          <td>(</td>
          <td class="paramtype">gbwt::GBWTBuilder *&#160;</td>
          <td class="paramname"><em>gbwt_builder</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Add transcript paths as threads in GBWT index. </p>

</div>
</div>
<a class="anchor" id="af921ea87f47cf03505aecd2286a2c05b"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::edit_graph </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> *&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const bool&#160;</td>
          <td class="paramname"><em>add_paths</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">
<p>Edits variation graph with transcript path splice-junctions and updates transcript path traversals to match the augmented graph. Optionally embed transcript paths in variation graph. </p>

</div>
</div>
<a class="anchor" id="a4808b2bd9c101b32da4e217abcac988b"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt; pair&lt; <a class="el" href="namespacevg.html#a8ae488d495e344881cbcf599ebdd2355">exon_nodes_t</a>, vector&lt; gbwt::size_type &gt; &gt; &gt; vg::Transcriptome::get_exon_haplotypes </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">vg::id_t</a>&#160;</td>
          <td class="paramname"><em>start_node</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const <a class="el" href="namespacevg.html#afaa15fd8bdfd688d48561e1b12aee821">vg::id_t</a>&#160;</td>
          <td class="paramname"><em>end_node</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const gbwt::GBWT &amp;&#160;</td>
          <td class="paramname"><em>haplotype_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const int32_t&#160;</td>
          <td class="paramname"><em>expected_length</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Extracts all unique haplotype paths between two nodes from a GBWT index and returns the resulting paths and the corresponding haplotype ids for each path. </p>

</div>
</div>
<a class="anchor" id="ac6aef1ffe0baa75f948859590e2ff33f"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">list&lt; <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a> &gt; vg::Transcriptome::project_transcript_embedded </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &amp;&#160;</td>
          <td class="paramname"><em>cur_transcript</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Projects transcripts onto embedded paths in a variation graph and returns resulting transcript paths. </p>

</div>
</div>
<a class="anchor" id="a5671f47e3233708fca381f1dce05b987"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">list&lt; <a class="el" href="structvg_1_1TranscriptPath.html">TranscriptPath</a> &gt; vg::Transcriptome::project_transcript_gbwt </td>
          <td>(</td>
          <td class="paramtype">const <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &amp;&#160;</td>
          <td class="paramname"><em>cur_transcript</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const gbwt::GBWT &amp;&#160;</td>
          <td class="paramname"><em>haplotype_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>mean_node_length</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Projects transcripts onto haplotypes in a GBWT index and returns resulting transcript paths. </p>

</div>
</div>
<a class="anchor" id="a963a41a941cd7ac2cf4857b110ed6e10"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::project_transcripts </td>
          <td>(</td>
          <td class="paramtype">const vector&lt; <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>transcripts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const gbwt::GBWT &amp;&#160;</td>
          <td class="paramname"><em>haplotype_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>mean_node_length</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Constructs transcript paths by projecting transcripts onto embedded paths in a variation graph and/or haplotypes in a GBWT index. </p>

</div>
</div>
<a class="anchor" id="a429527c91bc59be1efe68e1ca5984833"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::project_transcripts_callback </td>
          <td>(</td>
          <td class="paramtype">const int32_t&#160;</td>
          <td class="paramname"><em>thread_idx</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const vector&lt; <a class="el" href="structvg_1_1Transcript.html">Transcript</a> &gt; &amp;&#160;</td>
          <td class="paramname"><em>transcripts</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const gbwt::GBWT &amp;&#160;</td>
          <td class="paramname"><em>haplotype_index</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">const float&#160;</td>
          <td class="paramname"><em>mean_node_length</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Threaded transcript projecting. </p>

</div>
</div>
<a class="anchor" id="a130ef684755419776eaa83aae6a67bfa"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::reorder_exons </td>
          <td>(</td>
          <td class="paramtype"><a class="el" href="structvg_1_1Transcript.html">Transcript</a> *&#160;</td>
          <td class="paramname"><em>transcript</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p>Reverses exon order if the transcript is on the reverse strand and the exons are ordered in reverse. </p>

</div>
</div>
<a class="anchor" id="aa87de45115c76c3acfcb868391ae97bd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32_t vg::Transcriptome::size </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns number of transcript paths. </p>

</div>
</div>
<a class="anchor" id="a79f5fde9d7133ce828600482705c985e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">const vector&lt; <a class="el" href="structvg_1_1Path.html">Path</a> &gt; &amp; vg::Transcriptome::transcript_paths </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Returns transcript paths. </p>

</div>
</div>
<a class="anchor" id="a67d0d4546497376c5c31d248721f9c5e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::write_fasta_sequences </td>
          <td>(</td>
          <td class="paramtype">ostream *&#160;</td>
          <td class="paramname"><em>fasta_ostream</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype"><a class="el" href="classvg_1_1VG.html">VG</a> &amp;&#160;</td>
          <td class="paramname"><em>graph</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Writes transcript path sequences to a fasta file. </p>

</div>
</div>
<a class="anchor" id="ab76604980376271d6bbdffed718d8fe9"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void vg::Transcriptome::write_gam_alignments </td>
          <td>(</td>
          <td class="paramtype">ostream *&#160;</td>
          <td class="paramname"><em>gam_ostream</em></td><td>)</td>
          <td> const</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Writes transcript paths as alignments to a gam file. </p>

</div>
</div>
<h2 class="groupheader">Member Data Documentation</h2>
<a class="anchor" id="a127a44cf9f2a9d01064325e8e68cc948"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">vector&lt;<a class="el" href="structvg_1_1Path.html">Path</a>&gt; vg::Transcriptome::_transcriptome</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">
<p><a class="el" href="classvg_1_1Transcriptome.html">Transcriptome</a> represented by a set of transcript paths. TODO: Change to vector&lt;TranscriptPath&gt;. Current implementation decided in order for it to work easily with edit without excessive copying and moving of paths. </p>

</div>
</div>
<a class="anchor" id="aaef5359c4785d9e01e989a7875eed85e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Transcriptome::collapse_transcript_paths = true</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Collapse identical transcript paths. </p>

</div>
</div>
<a class="anchor" id="a78e782cd6f5d5c38730fbe54cfcbe3fc"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Transcriptome::filter_reference_transcript_paths = false</td>
        </tr>
      </table>
</div><div class="memdoc">

<p><a class="el" href="classvg_1_1Filter.html">Filter</a> transcript paths originating from a reference chromosome/contig. </p>

</div>
</div>
<a class="anchor" id="ac2261cd2e4236c02dbfded216253c923"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int32_t vg::Transcriptome::num_threads = 1</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Number of threads used for transcript path construction. </p>

</div>
</div>
<a class="anchor" id="a7f981a9d8172e931cc7c59b60de62caa"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">mutex vg::Transcriptome::trancriptome_mutex</td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Mutex used for adding transcript paths to transcriptome. </p>

</div>
</div>
<a class="anchor" id="a56a00bd350c02600e5474fc89791d5bd"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">string vg::Transcriptome::transcript_tag = &quot;transcript_id&quot;</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Attribute tag used to parse the transcript id/name in the gtf/gff file. </p>

</div>
</div>
<a class="anchor" id="afc815aefb14bad4cec9aeccdd9872652"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">bool vg::Transcriptome::use_embedded_paths = false</td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Use paths embedded in the graph for transcript path construction. </p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>src/<a class="el" href="transcriptome_8hpp.html">transcriptome.hpp</a></li>
<li>src/<a class="el" href="transcriptome_8cpp.html">transcriptome.cpp</a></li>
</ul>
</div><!-- contents -->
<!-- HTML footer for doxygen 1.8.9.1-->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.6
</small></address>
</body>
</html>
